File: crates/omnifs-release/src/semver.rs
use anyhow::{Result, bail};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct Version {
    pub major: u64,
    pub minor: u64,
    pub patch: u64,
    pub prerelease: Option<String>,
}

pub fn parse_version(input: &str) -> Result<Version> {
    let (core, prerelease) = match input.split_once('-') {
        Some((core, pre)) => (core, Some(pre.to_string())),
        None => (input, None),
    };
    let mut parts = core.split('.');
    let major = parts
        .next()
        .ok_or_else(|| anyhow::anyhow!("invalid version: {input}"))?
        .parse()?;
    let minor = parts
        .next()
        .ok_or_else(|| anyhow::anyhow!("invalid version: {input}"))?
        .parse()?;
    let patch = parts
        .next()
        .ok_or_else(|| anyhow::anyhow!("invalid version: {input}"))?
        .parse()?;
    if parts.next().is_some() {
        bail!("invalid version: {input}");
    }
    Ok(Version {
        major,
        minor,
        patch,
        prerelease,
    })
}

pub fn bump_patch(version: &str) -> Result<String> {
    let v = parse_version(version)?;
    Ok(format!("{}.{}.{}", v.major, v.minor, v.patch + 1))
}

File: scripts/lib/semver.ts
export type Version = {
  major: number;
  minor: number;
  patch: number;
  prerelease?: string;
};

export function parseVersion(input: string): Version {
  const [core, prerelease] = input.split("-", 2);
  const parts = core.split(".");
  if (parts.length !== 3) {
    throw new Error(`invalid version: ${input}`);
  }
  const [major, minor, patch] = parts.map((part) => Number(part));
  if (![major, minor, patch].every(Number.isSafeInteger)) {
    throw new Error(`invalid version: ${input}`);
  }
  return { major, minor, patch, prerelease };
}

export function formatVersion(version: Version): string {
  const core = `${version.major}.${version.minor}.${version.patch}`;
  return version.prerelease ? `${core}-${version.prerelease}` : core;
}

export function bumpPatch(version: Version): Version {
  return {
    major: version.major,
    minor: version.minor,
    patch: version.patch + 1,
  };
}

export type BumpLevel = "major" | "minor" | "patch";

export function bump(version: Version, level: BumpLevel): Version {
  switch (level) {
    case "major":
      return { major: version.major + 1, minor: 0, patch: 0 };
    case "minor":
      return { major: version.major, minor: version.minor + 1, patch: 0 };
    case "patch":
      return bumpPatch(version);
  }
}
